tool/tsh/common/db_exec.go

```diff
@@ -0,0 +1,647 @@
+	// Fetch.
+	dbs, err := c.getDatabases()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Execute  in parallel.
+	group, groupCtx := errgroup.WithContext(c.cf.Context)
+	group.SetLimit(c.cf.ParallelJobs)
+	for _, db := range dbs {
+		group.Go(func() error {
+			result := c.exec(groupCtx, db)
+			c.summary.add(result)
+			return nil
+		})
+	}
+
+	// Print summary.
+	defer func() {
+		switch {
+		case c.cf.OutputDir != "":
+			c.summary.printAndSave(c.cf.Stdout(), c.cf.OutputDir)
+		case len(dbs) > 1:
+			c.summary.print(c.cf.Stdout())
+		}
+	}()
+
+	return trace.Wrap(group.Wait())
+}
+
+func (c *databaseExecCommand) close() {
+	if err := c.client.close(); err != nil {
+		logger.WarnContext(c.cf.Context, "Failed to close client", "error", err)
+	}
+}
+
+func checkDatabaseExecInputFlags(cf *CLIConf) error {
+	// Pick an arbitrary number for max connections to avoid flooding the
+	// backend. The limit can be overwritten with the "TELEPORT_PARALLEL_JOBS"
+	// env var.
+	const maxParallelJobs = 10
+	if maxStr := os.Getenv(parallelJobsEnvVar); maxStr != "" {
+		if maxValue, err := strconv.Atoi(maxStr); err != nil || maxValue <= 0 {
+			return trace.BadParameter("environment variable %s must be a positive integer", parallelJobsEnvVar)
+		} else {
+			cf.ParallelJobs = maxValue
+		}
+	} else if cf.ParallelJobs < 1 || cf.ParallelJobs > maxParallelJobs {
+		return trace.BadParameter(`--parallel must be between 1 and %v.
+
+You may use TELEPORT_PARALLEL_JOBS to bypass limit. However, setting a very high
+value may overload your Teleport cluster. `, maxParallelJobs)
+	}
+
+	// Selection flags.
+	byNames := cf.DatabaseServices != ""
+	bySearch := cf.SearchKeywords != "" || cf.Labels != ""
+	switch {
+	case !byNames && !bySearch:
+		return trace.BadParameter("please provide one of --dbs, --labels, --search flags")
+	case byNames && bySearch:
+		return trace.BadParameter("--labels/--search flags cannot be used with --dbs flag")
+	}
+
+	// Logging.
+	if cf.ParallelJobs > 1 && cf.OutputDir == "" {
+		return trace.BadParameter("--output-dir must be set when executing concurrent connections")
+	}
+	if cf.OutputDir != "" && utils.FileExists(cf.OutputDir) {
+		return trace.BadParameter("directory %q already exists", cf.OutputDir)
+	}
+	return nil
+}
+
+func (c *databaseExecCommand) getDatabases() ([]types.Database, error) {
+	if c.cf.DatabaseServices != "" {
+		return c.getDatabasesByNames()
+	}
+	return c.searchDatabases()
+}
+
+func (c *databaseExecCommand) getDatabasesByNames() ([]types.Database, error) {
+	// Use a single predicate to search multiple names in one shot but batch 100
+	// names at a time. Extra validation will be performed afterward to ensure
+	// we fetched what we need.
+	fmt.Fprintln(c.cf.Stdout(), "Fetching databases by name ...")
+
+	// Trim spaces.
+	names := strings.Split(c.cf.DatabaseServices, ",")
+	for i := range names {
+		names[i] = strings.TrimSpace(names[i])
+	}
+	names = apiutils.Deduplicate(names)
```

Comment:
Nit: This looks like it could be a handy helper on the `c.cf` object. Or maybe a generic utility method? Split by comma, trim whitespaces and return deduped elements.